# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/ipc/win/handler/AccessibleHandlerControl.cpp
# Commit: 6658e0ae7e35
# Full Hash: 6658e0ae7e356f7b17896e1ee0316416fec2f2e0
# Author: James Teh <jteh@mozilla.com>
# Date: 2020-05-26 15:44:42
# Regressor Bug: 1640553
# File Overlap Count: 2
# Description:
#   Bug 1640553 part 1: Add the ability for AccessibleHandler to cache accessibles by id. r=eeejay
#   
#   This cache is cleared when other AccessibleHandler caches are invalidated; i.e. when an event is fired.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D76665
# ==============================================================================

diff -r 9ca07f6e0135 -r 6658e0ae7e35 accessible/ipc/win/handler/AccessibleHandlerControl.cpp
--- a/accessible/ipc/win/handler/AccessibleHandlerControl.cpp	Mon May 25 23:50:19 2020 +0000
+++ b/accessible/ipc/win/handler/AccessibleHandlerControl.cpp	Mon May 25 14:35:09 2020 +0000
@@ -118,6 +118,7 @@
 HRESULT
 AccessibleHandlerControl::Invalidate() {
   ++mCacheGen;
+  mAccessibleCache.clear();
   return S_OK;
 }
 
@@ -171,5 +172,23 @@
   return hr;
 }
 
+void AccessibleHandlerControl::CacheAccessible(long aUniqueId,
+                                               AccessibleHandler* aAccessible) {
+  MOZ_ASSERT(aUniqueId && aAccessible);
+  mAccessibleCache[aUniqueId] = aAccessible;
+}
+
+HRESULT AccessibleHandlerControl::GetCachedAccessible(
+    long aUniqueId, AccessibleHandler** aAccessible) {
+  MOZ_ASSERT(aUniqueId && aAccessible);
+  auto it = mAccessibleCache.find(aUniqueId);
+  if (it == mAccessibleCache.end()) {
+    return E_INVALIDARG;
+  }
+  RefPtr<AccessibleHandler> ref = it->second;
+  ref.forget(aAccessible);
+  return S_OK;
+}
+
 }  // namespace a11y
 }  // namespace mozilla